fcomm/src/lib.rs

```diff
@@ -896,32 +896,36 @@ impl Proof<Bls12> {
         let evaluation = &self.claim.evaluation().expect("expected evaluation claim");
 
         let input_io = {
-            let expr = s
-                .read(&evaluation.expr)
-                .ok_or_else(|| Error::VerificationError("failed to read expr".into()))?;
-            let env = s
-                .read(&evaluation.env)
-                .ok_or_else(|| Error::VerificationError("failed to read env".into()))?;
+            if let Ok(expr) = s.read(&evaluation.expr) {
+                if let Ok(env) = s.read(&evaluation.env) {
+                    // FIXME: We ignore cont and assume Outermost, since we can't read a Cont.
+                    let cont = s.intern_cont_outermost();
 
-            // FIXME: We ignore cont and assume Outermost, since we can't read a Cont.
-            let cont = s.intern_cont_outermost();
-
-            IO::<Scalar> { expr, env, cont }
+                    IO::<Scalar> { expr, env, cont }
+                } else {
+                    // FIXME: this fn should return an overarching
+                    // Error type where we can just return parser
+                    // errors as-is instead of hiding the root cause
+                    return Err(Error::VerificationError("failed to read env".into()));
+                }
+            } else {
+                return Err(Error::VerificationError("failed to read expr".into()));
+            }
         };
```

Comment:
The above nested expressions are extremely hard to read. I *think* this can be rewritten
```
        let input_io = {
            let expr = s
                .read(&evaluation.expr)
                .map_err(|_| Error::VerificationError("failed to read expr".into()))?;

            let env = s
                .read(&evaluation.env)
                .map_err(|_| Error::VerificationError("failed to read env".into()))?;

            // FIXME: We ignore cont and assume Outermost, since we can't read a Cont.
            let cont = s.intern_cont_outermost();

            IO::<Scalar> { expr, env, cont }
        };
```

Reply:
I can't recall my rationale for changing the structure here. I will try your way and if it works we can stick with it.